src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,753 @@
+    otherInfo.classInfo.emplace(otherInfo);
+    auto& classInfo = *otherInfo.classInfo;
+
+    // Update both groups to have the initial valid order. Their shapes still
+    // match.
+    classInfo.permute(otherInfo);
+    classInfo.permute(groupInfo);
+
+    // Insert `other` with its new shape. It may end up being joined to another
+    // existing equivalence class, in which case its class info will be cleared
+    // and `group` will subsequently be added to that same existing class, or
+    // alternatively it may be inserted as the representative element of a new
+    // class to which `group` will subsequently be joined.
+    shapesToUpdate.push_back(group);
+    shapesToUpdate.push_back(other);
+  }
+
+  std::vector<size_t>
+  getCanonicalPermutation(const std::vector<HeapType>& types) {
+    // The correctness of this part depends on some interesting properties of
+    // strongly connected graphs with ordered, directed edges. A permutation of
+    // the vertices in a graph is an isomorphism that produces an isomorphic
+    // graph. An automorphism is an isomorphism of a structure onto itself, so
+    // an automorphism of a graph is a permutation of the vertices that does not
+    // change the graph. Permutations can be described in terms of sets of
+    // cycles of elements.
+    //
+    // Theorem 1: All cycles in an automorphism of an SCC are the same size.
+    //
+    //     Proof: By contradiction. Assume there are two cycles of different
+    //     sizes. Because the SCC is fully connected, there is a path from an
+    //     element in the smaller of these cycles to an element in the larger.
+    //     This path must contain an edge between two vertices in cycles of
+    //     different sizes N and M such that N < M. Apply the automorphism N
+    //     times. The source of this edge has been cycled back to its original
+    //     index, but the destination is not yet back at its original index, so
+    //     the edge's destination index is different from its original
+    //     destination index and the permutation we applied was not an
+    //     automorphism after all.
+    //
+    // Corollary 1.1: No nontrivial automorphism of an SCC may have a stationary
+    // element, since either all the cycles have size 1 and the automorphism is
+    // trivial or all the cycles have some other size and there are no
+    // stationary elements.
+    //
+    // Corollary 1.2: No two isomorphic SCCs with the same first element are the
+    // same since no nontrivial automorphism can keep the first element
+    // stationary.
+    //
+    // Theorem 2: SCCs with initial elements that are not in an automorphism
+    // cycle with each other are not automorphic to each other.
+    //
+    //     Proof: By contradiction. Assume two such SCCs are automorphic to each
+    //     other. Then their initial elements must be in the same automorphism
+    //     cycle because they occupy the same index in two automorphic graphs.
```

Comment:
I'm not sure what you mean by "initial elements" here. This is the first time the term appears in the document that I can see.

I am guessing it means some prefix of the rec group? But that doesn't seem right, e.g.
```
(rec
  (type $a1 (struct (field (ref $a2) (ref $b1))))
  (type $b1 (struct (field (ref $b2) (ref $a1))))
  (type $a2 (struct (field (ref $a1) (ref $b2))))
  (type $b2 (struct (field (ref $b1) (ref $a2))))
)
```
We have the automorphism `($a1 $a2) ($b1 $b2)`, which does *not* keep the first two elements`$a1, $b1` in a cycle.

Reply:
By "initial element" I mean the element at index 0. In your example, applying the automorphism gives a graph with a different initial element, `$a2` instead of `$a1`. `$a1` and `$a2` are in the same cycle in some automorphism of the SCC.

Reply:
Ah, it's the plural that threw me:

> Then their initial elements

I get now that you mean the initial element in the two, so one element in each, but that was quite confusing...

Avoiding the plural here, or using the term "initial element" before, could be helpful here I think.

Reply:
See the last commit, which deletes theorem two and "inlines" the important takeaway into a later comment.